feat: функционал задачи управление календарем

SergeyZhitenev opened the pull request:
Прикладываю скриншоты работы приложения
![image](https://github.com/user-attachments/assets/16f685bf-c299-4d30-99c8-22af3cc813bb)
![image](https://github.com/user-attachments/assets/ad78da70-888b-45e8-b9fe-80c9024586b0)
![image](https://github.com/user-attachments/assets/4b735c85-13f4-4dc3-86d9-1b2432d09db9)
![image](https://github.com/user-attachments/assets/ab132bb2-ffa1-4de1-9e6e-ca8f5c618992)
![image](https://github.com/user-attachments/assets/d2f83b49-2e27-4312-8311-d5feefb062db)
![image](https://github.com/user-attachments/assets/3ba21538-ff1b-4b96-b2af-12e6d29069e2)
![image](https://github.com/user-attachments/assets/ce2eeebc-9a37-4f2c-8a9d-a4107516cbce)
Проверка функционала, что встречу может изменять только ее создатель 
![image](https://github.com/user-attachments/assets/445d86f3-1214-4f98-864f-5e524908d44f)
Обновление записи
![image](https://github.com/user-attachments/assets/95d42a18-148f-42dd-b1d7-25accf7256e2)
Удаление участника 
![image](https://github.com/user-attachments/assets/5b1c129d-8958-4d76-ad06-e87ed39124f7)
Отмена встречи
![image](https://github.com/user-attachments/assets/d813ae6a-3058-4d0d-a73e-1dd5175e2dec)
Несуществующий пользователь в качестве создателя
![image](https://github.com/user-attachments/assets/2e8db8cd-b667-40dd-9147-f8acf9b09d54)

SergeyZhitenev committed:
feat: функционал задачи управление календарем

SergeyZhitenev committed:
fix: delete testIfThrowsExceptionWhenDeletingVacancyWithNegativeId

The-Naku commented on src/main/java/faang/school/projectservice/filter/meetFilters/DateFilter.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public Stream<Meet> apply(Stream<Meet> requests, HttpServletRequest request) {
+        LocalDate ld = LocalDate.parse(request.getParameter("date"));
+        return requests.filter(meet -> meet.getCreatedAt().toLocalDate().equals(ld));
```
meet.getCreatedAt().toLocalDate().equals(ld)  нарушает инкапсуляцию.
куда элегантнее будет добавить в Meet  метод isSameCreated и передавать ему дату

The-Naku commented on src/main/java/faang/school/projectservice/filter/meetFilters/MeetFilter.java:
```diff
@@ -0,0 +1,12 @@
+import jakarta.servlet.http.HttpServletRequest;
+
+import java.util.stream.Stream;
+
```
реализации этого интерфейса лучше выносить в пакет impl

The-Naku commented on src/main/java/faang/school/projectservice/dto/meet/UpdateMeetDto.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.List;
+
+@Builder
+public record UpdateMeetDto(
```
а если при обновлении пользователь не хочет изменить какие то поля, а например только название или только описание? Мы получается его ограничиваем заставляя прописывать изменения для каждого поля, а если он не хочет меня поля, которые уже сущетсвуют и не изменяет их, то зачем нам передавать их в запросе и обьемом данных уменьшать трафик

SergeyZhitenev committed:
fix: error value
add HttpStatus.CREATED

The-Naku commented on src/main/java/faang/school/projectservice/service/MeetService.java:
```diff
@@ -0,0 +1,119 @@
+        meetValidator.validateMeetUpdating(meet);
+        meetValidator.validateThatRequestWasSentByTheCreator(meet, request);
+
+        meet.setTitle(updateMeetDto.title());
```
к комменту из updateDto о том, что пользователь не обязан изменять каждое поле. 
Также в библиотеке mapStruct есть удобный метод update, чтобы не городить сеттеры , который поддерживает частичное обновление

The-Naku commented on src/main/java/faang/school/projectservice/validator/MeetValidator.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public void validateThatRequestWasSentByTheCreator(Meet meet, HttpServletRequest request) {
+        long requesterId = Long.parseLong(request.getHeader("x-user-id"));
```
в лекции про feign client на  платформе есть про userContext, у которого есть метод getUserId, который позволяет нам получить id пользователя из хедера запроса и не нужно будет передавать в контроллер обьект с данными самого запроса и самому вытаскивать значение хедера, ибо эта функция у нас уже реализована в проекте. Ты молодец, что разобрался как это делать, однако думаю не стоит расширять кодовую база кодом с одинаковой функциональностью  

и твоя валидация относится к бизнес логике скорее  и логичнее ее все же оставить в сервисе, мне кажется 
это относится и к методам ниже

The-Naku commented on src/main/java/faang/school/projectservice/service/MeetService.java:
```diff
@@ -0,0 +1,119 @@
+        meetValidator.validateThatRequestWasSentByTheCreator(meet, request);
+        meetValidator.validateParticipants(userId, meet);
+
+        meet.getUserIds().remove(userId);
```
также тут нарушается инкапсуляция , было бы элегантнее создать метод с говорящим названием  в сущности по удалению участника из списка участников и вызывать его

The-Naku commented on src/main/java/faang/school/projectservice/service/MeetService.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    @Transactional
+    public List<MeetDto> getMeets(HttpServletRequest request) {
```
как будто возможно лучше было бы создать дто с полями всех фильтров и получать эти фильтры из тела запроса от клиента, было бы нагляднее какие и сколько фильтров у нас вообще есть
Также есть вероятность, что кол-во фильтров может сильно увеличится и их клиент  захочет применить все одновременно и тогда URL может очень сильно разрастись 

Это мои мысли на этот счет, и возможно твой вариант тоже хорош
Интересно уточнить у Ильи на этот счет

SergeyZhitenev committed:
feat: изменена проверка отправителя запроса
методы валидации перенесены в сервис
исправлено нарушение инкапусляции в сервисе
Добавлены @NotNull проверки в дто

The-Naku commented on src/main/java/faang/school/projectservice/controller/MeetController.java:
```diff
@@ -0,0 +1,63 @@
+
+    @PostMapping("/create")
+    @ResponseStatus(HttpStatus.CREATED)
+    public MeetDto create(@Validated @RequestBody MeetDto meetDto) {
```
почитай о том как правильно проектировать ендпоинты  по ресту. слов действий (за редким исключением) не должно быть в URI. когда у тебя приходит запрос с post или delete не нужно указывать, что это происходит в URI так как у нас запрос приходит с конкретным типом действия. 
путь должен обозначать ресурс, а не действие. действие должно быть выражено через http методы

SergeyZhitenev committed:
fix: измененые url в контроллерах
Убран отдельный метод на отмену встречи ( отменять через обновление )

ipasynkov-epam commented:
![image](https://github.com/user-attachments/assets/f147f4af-6148-4d18-94ca-338d854b7431)
Сейчас нужно создать ПР в другую ветку, в мастер мержить не получится)
